internal/controllers/uiplugin_controller.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright © 2023 - 2024 SUSE LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// UIPluginReconciler reconciles a UIPlugin object.
+type UIPluginReconciler struct {
+	client.Client
+	*runtime.Scheme
+	UncachedClient client.Client
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *UIPluginReconciler) SetupWithManager(_ context.Context, mgr ctrl.Manager, _ controller.Options) error {
+	uiPlugin := &metav1.PartialObjectMetadata{}
+	uiPlugin.SetGroupVersionKind(schema.GroupVersionKind{
```

Comment:
should this be a part of the rancher API wrapper?

Reply:
It can be, but this controller does not set any spec fields, only moves the UI plugin to the `cattle-ui-system-namespace`, once this namespace is created. It may also be used later to not move and simply delete `UIPlugin` if it interferes with existing one in the namespace already. Helm has problems with cleaning up or replacing resources if they weren’t created by it, so a controller is better suited to accommodate this custom logic.